### DISCUSSION TITLE
Q. about dark theme files

### DISCUSSION THREAD

**Discussion by klaus101** [doublecmd/doublecmd#884] [General]
Hopefully it's not out of place here to ask 2 questions that are not directly related to DC, but some files creating the dark theme.
Occasionally i was made aware of them, and purely by interest i tried them in and own app, exclusively used for private needs. Hopefully i'd be allowed to.
i tried them, using Lazarus, and was extremely excited. Excellent! Forms and controls were shown really great (Win 10 x64)

The only caveat i encountered had been the column headers of the ListView, which still appeared white.
So i'd be very interested to know: are the plans, or is it foreseen, that uwin32widgetsetdark does cover the listview column headers too?

Secondly: actually i handle the code branching (dark or not?) via a $DEFINE and $IFDEFs. 
Ideally one wants that the app reads an inifile key once that controls the app in that way that, if the key is set to ‘no-dark’,
udarkstyle as well as uwin32widgetsetdark might exit at an early beginning. Is it possible by concept, something analogue to the EarlyConfig, but for both files, where the info is passed from  the mainform? I'm asking because i don't know if there's already a hook which i simply didn't see.

**Comment by alexx2000** [comment]
1. DC does not use TListView that's why my units don't make it dark. Technically I guess it can be done.
2. My implementation requires that dark mode should be enabled before main form creation.

**Comment by klaus101** [comment]
@alexx2000 
1) Assumed so when i saw that DC doesn't use the listview.
   Maybe you change your mind some day and integrate the listview columnheaders in the widegeset dark file anyway? .... Wpuld be a very special holiday!

2) I had the "use" first, in a test project, within the project file (where an inikey could be read too). 
   When applying it on a my main app (used as a major part of my personal working environmnt),
   i included the files from within the mainform, from where the dark mode inherits to the subforms.
   Works perfect and excellent (except the column headers") Amazing! My deep respect!
   But actually there is no mechanism to pass info to the files (for to optionally not to begin the work)?
   Maybe i try it by applying s global variable (never used those, so i don't know yet if it's a good idea).

**Reply by klaus101** [reply]
Alex, i'd have a question to you.
Did some trying and could achieve at last the dark listview column header background
with a very few lines within uwin32widgetsetdark.pas (see below). The key is: 'ItemsView::Header'.
Works just as expected (including the sort indicator, if any).

But the font color stays dark, unreadable so.
The challenge would be to modify the font color within uwin32widgetsetdark.pas, so that not any using class depends on own subclassing, ownerdraw & Co.).
As you know the dark theme files best: do you see an already useable hook for this, to change the font color, when the handle of the control is already available? Would be the last missing piece within the listview puzzle.
![grafik](https://user-images.githubusercontent.com/31130146/229163671-899fcc27-2a5b-4248-87ef-3dada0705063.png)

I add a small test project (Lazarus) what i did.
[test_ListView_Header_Darkmode_2.zip](https://github.com/doublecmd/doublecmd/files/11123406/test_ListView_Header_Darkmode_2.zip)

```
function InterceptOpenThemeData(hwnd: hwnd; pszClassList: LPCWSTR): hTheme; stdcall;
var
  P: LONG_PTR;
  hColHdr: HWND;  // <- added
....
// Somewhere within the classlist evaluation if-else:
....
  else if lstrcmpiW(pszClassList, 'ListView') = 0 then
  begin
     hColHdr := Listview_GetHeader(hwnd);    // Listview_GetHeader in: CommCtrl. HWND in lcltype
     if hColHdr <> 0 then begin
       AllowDarkModeForWindow(hColHdr, True);
       pszClassList:= PWideChar('ItemsView::Header');
       SetWindowTheme(hColHdr, 'ItemsView', NIL);    // Please note that the ListView itself stays intentionally at theme 'DarkMode_Explorer'
       SendMessageW(hColHdr, WM_THEMECHANGED, 0, 0);
     end;
  end
  else if  ............ and so on
  

```

**Comment by klaus101** [comment]
About the ListView, finally i could find a solution by myself.
Alex, could you please tell me in short what you think about the solution?

[__test_ListView_Header_Darkmode_4_ok.zip](https://github.com/doublecmd/doublecmd/files/11130568/__test_ListView_Header_Darkmode_4_ok.zip)

![LostView_Dark_OK](https://user-images.githubusercontent.com/31130146/229313480-e09227cc-375c-4623-b929-3ef3088d72a9.jpg)
